### ISSUE TITLE
No valid configuration found

### ISSUE THREAD

**Issue by Earnestly** [AdaCore/gprbuild#41]
*Edit: Submitting issue for public visibility.*

When using `env -i HOME="$HOME" PATH=/bin gprbuild ...` where gprconfig data is located in `/usr/share/gprconfig`, `gprbuild` will fail as it searches under `/share` which does not exist.

Unbelievably, changing `PATH` to `PATH=/usr/bin` allows `gprbuild` to correctly locate these files, but will remain broken if `PATH=/bin:/usr/bin` is used.

`gprbuild` provides an undocumented (only present in `--help`) option for allowing one to specify a location which works around this issue by using `--db /usr/share/gprconfig`.

However `gprinstall` lacks this option, or another equivalent option, and so remains dependant on `PATH`.

**Comment by Earnestly** [comment]
Also still cannot specify a specific C compiler via the `gprbuild` command-line.

**Comment by Earnestly** [comment]
Additionally, something in my environment is allowing xmlada to be built with `static-pic` and `relocatable` build types that fails when using a minimal environment as shown in the first message here.

I have not yet figured out what my environment contains that would lead to this since nothing there is related to Ada, gcc, etc.

**Comment by Earnestly** [comment]
`raised ADA.IO_EXCEPTIONS.END_ERROR : a-tigeli.adb:96`

Just for fun, this will randomly happen when using `gprinstall`.  So far the workaround is to simply try again until it doesn't happen.

**Comment by t-14** [comment]
> When using env -i HOME="$HOME" PATH=/bin gprbuild ... where gprconfig data is located in /usr/share/gprconfig, gprbuild will fail as it searches under /share which does not exist.

Is /bin a symlink to /usr/bin on your system or these are two distinct directories?

> Also still cannot specify a specific C compiler via the gprbuild command-line.

As explained before, to control the toolchains you need to run gprconfig first then pass cgpr file to gprbuild.

> raised ADA.IO_EXCEPTIONS.END_ERROR : a-tigeli.adb:96

> Just for fun, this will randomly happen when using gprinstall. So far the workaround is to simply try again until it doesn't happen.

This suggests a race condition resulting in unexpected empty stream in get_line. Are you running several gprinstall in parallel by any chance (e.g. by passing -j to make?)

**Comment by Earnestly** [comment]
Yup, `/bin` is a symlink to `/usr/bin`.  It's fairly common these days due to https://www.freedesktop.org/wiki/Software/systemd/TheCaseForTheUsrMerge/ and https://fedoraproject.org/wiki/Features/UsrMove

For all build types, `gprinstall` uses `-XPROCESSORS=0` so I don't believe it could be related to that.  `gprbuild` itself is using `-j0` as well.

**Comment by t-14** [comment]
> Yes, /bin is a symlink to /usr/bin

Understood. I confirm that we make no attempt at the moment to resolve symlinks in this case. We probably should since that's what gcc does. We'll look into that.

> For all build types, gprinstall uses -XPROCESSORS=0. gprbuild itself is using -j0 as well. However I do use MAKEFLAGS=-j$(nproc) but I wouldn't know how that could cause an issue with gprinstall itself.

If you run "make -j0", it will potentially launch several make targets in parallel. You can end up with gprinstall running in parallel with gprbuild whose artifacts gprinstall is supposed to install, or with several gprinstall running in parallel trying to update the same installation root. Neither of this is supported.

In summary, if makefile is used to drive gpr* tools, make targets should always be sequential. Speaking of which, we will add .NOTPARALLEL target at least to gprbuild's own makefile to mitigate this.

> In my opinion, generating a *.cgpr is far too involved to merely change one detail when one is used to simply supplying a different value to the CC enviornment.

But that's what gprbuild will have to do in autoconf mode anyway. No way around it, since changing a toolchain can change its characteristics and affect the resulting build process (like what options are supported, etc). The only question is if we want to intervene in this process by taking control of gprconfig command.
